File: app/src/main/java/cn/chenyuanming/gankmeizhi/adapter/ArticleViewAdapter.java
package cn.chenyuanming.gankmeizhi.adapter;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.List;
import java.util.TreeSet;

import butterknife.Bind;
import butterknife.ButterKnife;
import cn.chenyuanming.gankmeizhi.R;
import cn.chenyuanming.gankmeizhi.activity.ShowBigImageActivity;
import cn.chenyuanming.gankmeizhi.activity.WebViewActivity;
import cn.chenyuanming.gankmeizhi.beans.FavoriteBean;
import cn.chenyuanming.gankmeizhi.beans.GoodsBean;
import cn.chenyuanming.gankmeizhi.beans.ReadArticles;
import cn.chenyuanming.gankmeizhi.utils.DbHelper;

/**
 * Created by Administrator on 2016/1/28.
 */
public class ArticleViewAdapter extends RecyclerView.Adapter<ArticleViewAdapter.ViewHolder> {

    private List<GoodsBean.Results> mDatas = new ArrayList<>();

    FavoriteBean favorite = DbHelper.getHelper().getData(FavoriteBean.class).get(0);
    ReadArticles readArticles = DbHelper.getHelper().getData(ReadArticles.class).get(0);

    public static class ViewHolder extends RecyclerView.ViewHolder {
        @Bind(R.id.iv_meizhi)
        ImageView iv_meizhi;
        @Bind(R.id.tv_type)
        TextView tv_type;
        @Bind(R.id.tv_title)
        TextView tv_title;
        @Bind(R.id.tv_author)
        TextView tv_author;
        @Bind(R.id.iv_share)
        ImageView iv_share;
        @Bind(R.id.iv_favorite)
        ImageView ivFavorite;


        public ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

    public String getValueAt(int position) {
        return mDatas.get(position).desc;
    }

    Context context;
    int fragType;

    public ArticleViewAdapter(Context context, List<GoodsBean.Results> items, int fragType) {
        this.context = context;
        if (items != null) {
            mDatas.addAll(items);
        }
        this.fragType = fragType;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_article, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        GoodsBean.Results data = mDatas.get(position);

        holder.tv_title.setText(data.desc);
        holder.tv_author.setText("推荐 by @" + data.who);
        holder.tv_type.setText("#" + data.type + " ");
        if (readArticles.articles.contains(data.objectId)) {
            holder.tv_title.setTextColor(context.getResources().getColor(R.color.lightBlack));
        } else {
            holder.tv_title.setTextColor(context.getResources().getColor(R.color.black));
        }
        if (data.type.equals("福利")) {
            holder.iv_meizhi.setVisibility(View.VISIBLE);
            Glide.with(context).load(data.url).into(holder.iv_meizhi);
            holder.iv_meizhi.setOnClickListener((v) -> {
                Intent intent = new Intent(context, ShowBigImageActivity.class);
                intent.putExtra("data", mDatas.get(position));
                context.startActivity(intent);
            });
        } else {
            holder.iv_meizhi.setVisibility(View.GONE);
            holder.tv_title.setOnClickListener(v -> {
                holder.tv_title.setTextColor(context.getResources().getColor(R.color.lightBlack));
                Intent intent = new Intent(context, WebViewActivity.class);
                intent.putExtra("url", mDatas.get(position).url);
                intent.putExtra("objectId", mDatas.get(position).objectId);
                context.startActivity(intent);
                readArticles.articles.add(data.objectId);
                DbHelper.getHelper().getLiteOrm().save(readArticles);
            });
        }

        setFavoriteIcon(holder.ivFavorite, favorite.favorites, data.objectId);
        holder.ivFavorite.setOnClickListener(v -> {
            onFavoriteClicked(holder.ivFavorite, favorite.favorites, data.objectId);
            DbHelper.getHelper().getLiteOrm().save(favorite);
        });
    }

    private void onFavoriteClicked(ImageView ivFavorite, TreeSet<String> favorites, String objectId) {
        if (favorites.contains(objectId)) {
            favorites.remove(objectId);
        } else {
            favorites.add(objectId);
        }
        setFavoriteIcon(ivFavorite, favorites, objectId);
    }

    private void setFavoriteIcon(ImageView ivFavorite, TreeSet<String> favorites, String objectId) {
        Drawable drawable = ivFavorite.getDrawable();
        if (favorites.contains(objectId)) {
            drawable.setColorFilter(Color.parseColor("#ff0000"), PorterDuff.Mode.SRC_IN);
            ivFavorite.setImageDrawable(drawable);
        } else {
            ivFavorite.setImageResource(R.drawable.ic_favorite);
        }
    }

    @Override
    public int getItemCount() {
        return mDatas.size();
    }

    public List<GoodsBean.Results> getDatas() {
        return mDatas;
    }

}

File: app/src/main/java/cn/chenyuanming/gankmeizhi/adapter/ArticleViewAdapter.kt
package cn.chenyuanming.gankmeizhi.adapter

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import cn.chenyuanming.gankmeizhi.R
import cn.chenyuanming.gankmeizhi.activity.ShowBigImageActivity
import cn.chenyuanming.gankmeizhi.activity.WebViewActivity
import cn.chenyuanming.gankmeizhi.application.GankApplication.Companion.context
import cn.chenyuanming.gankmeizhi.beans.CommonGoodsBean
import cn.chenyuanming.gankmeizhi.beans.db.FavoriteBean
import cn.chenyuanming.gankmeizhi.beans.db.ReadArticles
import cn.chenyuanming.gankmeizhi.utils.DbHelper
import cn.chenyuanming.gankmeizhi.utils.ShareUtils
import cn.chenyuanming.gankmeizhi.utils.TimeHelper
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.activity_webview.view.*
import kotlinx.android.synthetic.main.item_article.view.*
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by Chen Yuanming on 2016/1/28.
 */
class ArticleViewAdapter(items: List<CommonGoodsBean.Results>?, internal var fragType: Int) : RecyclerView.Adapter<ArticleViewAdapter.ViewHolder>() {

    private val mDatas = ArrayList<CommonGoodsBean.Results>()

    internal var favorite = DbHelper.getHelper().getData(FavoriteBean::class.java)[0]
    internal var readArticles = DbHelper.getHelper().getData(ReadArticles::class.java)[0]
    internal var currDate = Date()

    val datas: MutableList<CommonGoodsBean.Results>
        get() = mDatas

    class ViewHolder(internal var mainView: View) : RecyclerView.ViewHolder(mainView) {

        init {
            context = mainView.context
            defaultShare = mainView.ivShare.drawable
            defaultShare.setColorFilter(Color.parseColor("#bfc8d6"), PorterDuff.Mode.SRC_IN)
            mainView.ivShare.setImageDrawable(defaultShare)
        }
    }

    fun getValueAt(position: Int): String? {
        return mDatas[position].desc
    }

    init {
        if (items != null) {
            mDatas.addAll(items)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_article, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val data = mDatas[position]

        holder.mainView.tvTitle.text = data.desc

        try {
            val publishDate = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").parse(data.publishedAt)
            holder.mainView.tvTime.text = TimeHelper.getTime(publishDate.time)
        } catch (e: ParseException) {
            holder.mainView.tvTime.text = data.updatedAt.substring(0, data.updatedAt.indexOf("T"))
            e.printStackTrace()
        }

        holder.mainView.tvAuthor.text = "by @" + data.who
        holder.mainView.tvType.text = data.type
        if (readArticles.articles.contains(data.url)) {
            holder.mainView.tvTitle.setTextColor(context.resources.getColor(R.color.lightBlack))
        } else {
            holder.mainView.tvTitle.setTextColor(context.resources.getColor(R.color.black))
        }


        if (data.type == "福利") {
            holder.mainView.tvTitle.visibility = View.GONE
            holder.mainView.ivMeizhi.visibility = View.VISIBLE
            Glide.with(context).load(data.url).into(holder.mainView.ivMeizhi)
            holder.mainView.setOnClickListener { v ->
                val intent = Intent(context, ShowBigImageActivity::class.java)
                intent.putExtra("data", mDatas[position])
                context.startActivity(intent)
            }
        } else {
            holder.mainView.tvTitle.visibility = View.VISIBLE
            holder.mainView.ivMeizhi.visibility = View.GONE
            holder.mainView.setOnClickListener { v ->
                holder.mainView.tvTitle.setTextColor(context.resources.getColor(R.color.lightBlack))
                val intent = Intent(context, WebViewActivity::class.java)
                intent.putExtra("url", mDatas[position].url)
                context.startActivity(intent)
                readArticles.articles.add(data.url)
                DbHelper.getHelper().liteOrm.save(readArticles)
            }
        }

        setFavoriteIcon(holder.mainView.ivFavorite, favorite.favorites, data.url)
        holder.mainView.ivFavorite.setOnClickListener { v ->
            onFavoriteClicked(holder.mainView.ivFavorite, favorite.favorites, data.url)
            DbHelper.getHelper().liteOrm.save(favorite)
        }

        holder.mainView.ivShare.setOnClickListener { v -> ShareUtils.share(context, data.desc + data.url) }

//        when (data.type) {
//            "Android" -> holder.mainView.tvType.setBackgroundResource(R.drawable.shapeType_android)
//            "iOS" -> holder.mainView.tvType.setBackgroundResource(R.drawable.shapeType_ios)
//            "拓展资源", "App" -> holder.mainView.tvType.setBackgroundResource(R.drawable.shapeType_extend)
//            "休息视频" -> holder.mainView.tvType.setBackgroundResource(R.drawable.shapeType_relax)
//            else -> holder.mainView.tvType.setBackgroundResource(R.drawable.shapeType_relax)
//        }
    }

    private fun initWebview(webView: WebView) {
        webView.webChromeClient = WebChromeClient()
        webView.webViewClient = WebViewClient()
        webView.settings.builtInZoomControls = true
        webView.settings.javaScriptEnabled = true
        webView.settings.domStorageEnabled = true
        webView.settings.databaseEnabled = true
        webView.settings.cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK
        webView.settings.setAppCacheEnabled(true)

        webView.requestFocus()
        webView.settings.allowFileAccess
    }

    private fun onFavoriteClicked(ivFavorite: ImageView, favorites: TreeSet<String>, objectId: String) {
        if (favorites.contains(objectId)) {
            favorites.remove(objectId)
        } else {
            favorites.add(objectId)
        }
        setFavoriteIcon(ivFavorite, favorites, objectId)
    }

    private fun setFavoriteIcon(ivFavorite: ImageView, favorites: TreeSet<String>, objectId: String) {
        val drawable = ivFavorite.drawable
        if (favorites.contains(objectId)) {
            drawable.setColorFilter(Color.parseColor("#ff0000"), PorterDuff.Mode.SRC_IN)
            ivFavorite.setImageDrawable(drawable)
        } else {
            drawable.setColorFilter(Color.parseColor("#bfc8d6"), PorterDuff.Mode.SRC_IN)
            ivFavorite.setImageDrawable(drawable)
        }
    }

    override fun getItemCount(): Int {
        return mDatas.size
    }

    companion object {
        lateinit var defaultShare: Drawable
        lateinit var context: Context
    }

}
